ambry-store/src/main/java/com.github.ambry.store/IndexValue.java

```diff
@@ -13,114 +13,322 @@
+  private final long operationTimeInMs;
+  private final short serviceId;
+  private final short containerId;
+  private final short version;
 
-  IndexValue(String logSegmentName, ByteBuffer value) {
-    if (value.capacity() != Index_Value_Size_In_Bytes) {
-      throw new IllegalArgumentException("Invalid buffer size");
+  /**
+   * Constructs the {@link IndexValue} with the passed in {@link ByteBuffer} and the given {@code version}
+   * @param logSegmentName the log segment name to be used to construct the offset
+   * @param value the {@link ByteBuffer} representation of the {@link IndexValue}
+   * @param version the version of the {@link PersistentIndex}
+   */
+  IndexValue(String logSegmentName, ByteBuffer value, short version) {
+    this.version = version;
+    switch (version) {
+      case PersistentIndex.VERSION_0:
+        if (value.capacity() != INDEX_VALUE_SIZE_IN_BYTES_V0) {
+          throw new IllegalArgumentException("Invalid buffer size for version 0");
+        }
+        size = value.getLong();
+        offset = new Offset(logSegmentName, value.getLong());
+        flags = value.get();
+        expiresAtMs = value.getLong();
+        originalMessageOffset = value.getLong();
+        operationTimeInMs = (int) Utils.Infinite_Time;
+        serviceId = SERVICE_CONTAINER_ID_DEFAULT_VALUE;
+        containerId = SERVICE_CONTAINER_ID_DEFAULT_VALUE;
+        break;
+      case PersistentIndex.VERSION_1:
+        if (value.capacity() != INDEX_VALUE_SIZE_IN_BYTES_V1) {
+          throw new IllegalArgumentException("Invalid buffer size for version 1");
+        }
+        size = value.getLong();
+        offset = new Offset(logSegmentName, value.getLong());
+        flags = value.get();
+        int expiresAt = value.getInt();
+        expiresAtMs = expiresAt != Utils.Infinite_Time ? (expiresAt * Time.MsPerSec) : Utils.Infinite_Time;
+        originalMessageOffset = value.getLong();
+        int operationTimeInSecs = value.getInt();
+        operationTimeInMs =
+            operationTimeInSecs != Utils.Infinite_Time ? (operationTimeInSecs * Time.MsPerSec) : Utils.Infinite_Time;
+        serviceId = value.getShort();
+        containerId = value.getShort();
+        break;
+      default:
+        throw new IllegalArgumentException("Unsupported version " + version + " passed in for IndexValue ");
     }
-    this.value = value;
-    offset = new Offset(logSegmentName, value.getLong(Blob_Size_In_Bytes));
   }
 
-  IndexValue(long size, Offset offset, byte flags, long expiresAtMs) {
-    this(size, offset, flags, expiresAtMs, offset.getOffset());
+  /**
+   * Constructs IndexValue based on the args passed
+   * @param size the size of the blob that this index value refers to
+   * @param offset the {@link Offset} in the {@link Log} where the blob that this index value refers to resides
+   * @param expiresAtMs the expiration time in ms at which the blob expires
+   */
+  IndexValue(long size, Offset offset, long expiresAtMs) {
```

Comment:
This constructor is being used by the PersistentIndex and most of PROD code for now